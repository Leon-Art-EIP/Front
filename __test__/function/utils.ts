import type { ReactTestRendererJSON } from "react-test-renderer";
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from "react-test-renderer";
import { expect } from "vitest";

function toJson(component: renderer.ReactTestRenderer): ReactTestRendererJSON {
  const result = component.toJSON();
  expect(result).toBeDefined();

  return result as ReactTestRendererJSON;
}

export function testSnapshot(element: JSX.Element): void {
  const component = renderer.create(element);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
}
