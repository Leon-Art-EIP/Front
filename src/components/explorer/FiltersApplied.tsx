import React from 'react';

export default function FiltersApplied(): JSX.Element {
  return (
    <div>
      <h1>FiltersApplied</h1>
    </div>
  );
}