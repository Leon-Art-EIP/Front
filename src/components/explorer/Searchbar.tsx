import ClearIcon from "@mui/icons-material/Clear";
import SearchIcon from "@mui/icons-material/Search";
import { useState } from "react";

interface SearchbarProps {
  handleSearchTerm: (searchTerm: string) => void;
}

export default function Searchbar(props: SearchbarProps): JSX.Element {
  const [searchTerm, setSearchTerm] = useState<string>("");

  function onSearchTerm(e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();
    props.handleSearchTerm(searchTerm);
  }

  return (
    <form className="relative w-full" onSubmit={onSearchTerm}>
      <input
        className="w-full h-14 bg-secondary rounded-3xl pl-16 focus:shadow-md focus:outline-none focus:ring-1 focus:ring-tertiary transition duration-300 ease-in-out text-tertiary"
        placeholder="Rechercher une oeuvre ou un utilisateur..."
        onChange={(e) => setSearchTerm(e.target.value)}
        value={searchTerm}
      />
      <button
        type="submit"
        className="p-2 absolute top-2 left-4 hover:opacity-75 rounded-xl transition hover:bg-secondary hover:shadow-lg duration-300 ease-in-out"
      >
        <SearchIcon className={`text-tertiary`} />
      </button>
      {searchTerm && (
        <button className="absolute top-2 right-2">
          <ClearIcon className="text-primary" />
        </button>
      )}
    </form>
  );
}
