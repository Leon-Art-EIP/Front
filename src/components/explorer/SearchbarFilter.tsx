import React from 'react';

export default function SearchbarFilter(): JSX.Element {
  return (
    <div>
      <h1>SearchbarFilter</h1>
    </div>
  );
}