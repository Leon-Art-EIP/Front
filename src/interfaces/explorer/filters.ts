export interface IFilters {
  searchTerm: string;
  artType: string;
  priceRange: string;
  isForSale: boolean;
  sort: string;
  artPage: number;
  artLimit: number;
  artistPage: number;
  artistLimit: number;
}