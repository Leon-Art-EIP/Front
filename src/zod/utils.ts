export const getValueOrUndefined = (arg: unknown) =>
  arg === "" || arg === undefined || arg === null ? undefined : arg;
