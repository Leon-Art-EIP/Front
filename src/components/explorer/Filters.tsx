import React, { useState, useRef, useEffect } from "react";
import FilterAltIcon from "@mui/icons-material/FilterAlt";
import SortIcon from "@mui/icons-material/Sort";
import CheckIcon from "@mui/icons-material/Check";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";
import ExpandLessIcon from "@mui/icons-material/ExpandLess";
import { artTypeFilters, ArtTypeFilter } from "../../configs/explorer/artTypes";
import { IFilters } from "../../interfaces/explorer/filters";

interface FiltersProps {
  handleFilters: (filters: IFilters) => void;
  filters: IFilters;
}

export default function Filters(props: FiltersProps): JSX.Element {
  const [collapseFilters, setCollapseFilters] = useState<boolean>(false);
  const [collapseSort, setCollapseSort] = useState<boolean>(false);

  const filterRef = useRef<HTMLDivElement>(null);
  const sortRef = useRef<HTMLDivElement>(null);
  const filterButtonRef = useRef<HTMLButtonElement>(null);
  const sortButtonRef = useRef<HTMLButtonElement>(null);

  const [artTypes, setArtTypes] = useState<ArtTypeFilter[]>(artTypeFilters);

  function onCollapseFilter(event: React.MouseEvent) {
    event.stopPropagation();
    setCollapseFilters(!collapseFilters);
    if (collapseSort) setCollapseSort(false);
  }

  function onCollapseSort(event: React.MouseEvent) {
    event.stopPropagation();
    setCollapseSort(!collapseSort);
    if (collapseFilters) setCollapseFilters(false);
  }

  useEffect(() => {
    function handleClickOutside(event: MouseEvent) {
      const target = event.target as Node;

      // Vérifiez si le clic est dans les références des boutons ou des dropdowns
      if (
        (filterButtonRef.current && filterButtonRef.current.contains(target)) ||
        (sortButtonRef.current && sortButtonRef.current.contains(target)) ||
        (filterRef.current && filterRef.current.contains(target)) ||
        (sortRef.current && sortRef.current.contains(target))
      ) {
        return;
      }

      setCollapseFilters(false);
      setCollapseSort(false);
    }

    document.addEventListener("mousedown", handleClickOutside);

    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, []);

  function toggleArtCategory(index: number) {
    const newArtTypes = [...artTypes];
    newArtTypes[index].collapsed = !newArtTypes[index].collapsed;
    setArtTypes(newArtTypes);
  }

  function selectArtType(type: string) {
    const newArtTypes = [...artTypes];
    newArtTypes.forEach((artType) => {
      artType.types.forEach((artTypeType) => {
        if (artTypeType.type === type) artTypeType.selected = !artTypeType.selected;
      });
    });
    setArtTypes(newArtTypes);
  }

  return (
    <div className="flex flex-row gap-4">
      <div className="relative flex flex-col gap-4">
        <button
          type="button"
          ref={filterButtonRef}
          className="py-3 px-4 flex flex-row gap-2 items-center rounded-xl transition hover:bg-gray-200 hover:shadow-lg duration-300 ease-in-out"
          onClick={(event) => onCollapseFilter(event)}
        >
          <span className="text-lg font-medium">Filtres</span>
          <FilterAltIcon />
        </button>
        {collapseFilters && (
          <div
            ref={filterRef}
            className="absolute top-[110%] right-0 w-[320px] h-[450px] overflow-y-auto z-10 bg-gray-100 shadow-lg rounded-xl p-4"
          >
            <div className="flex flex-col gap-2">
              {artTypes.map((artType, index) => (
                <div key={index}>
                  <button
                    type="button"
                    className="py-2 w-full flex flex-row justify-between items-center px-4 rounded-xl hover:bg-gray-200 hover:shadow-lg duration-300 ease-in-out"
                    onClick={() => toggleArtCategory(index)}
                  >
                    {artType.category}
                    {artType.collapsed ? <ExpandMoreIcon /> : <ExpandLessIcon />}
                  </button>
                  {artType.collapsed && (
                    <div className="flex flex-col gap-2 pl-4">
                      {artType.types.map((type, index) => (
                        <button
                          key={index}
                          type="button"
                          className={`py-2 w-full flex flex-row justify-between items-center px-4 rounded-xl hover:bg-gray-200 hover:shadow-lg duration-300 ease-in-out ${
                            type.selected ? "bg-gray-200" : ""
                          }`}
                          onClick={() => selectArtType(type.type)}
                        >
                          {type.type}
                          {type.selected && <CheckIcon className="text-green-300" />}
                        </button>
                      ))}
                    </div>
                  )}
                </div>
              ))}
            </div>
          </div>
        )}
      </div>
      <div className="relative flex flex-col gap-4">
        <button
          type="button"
          id="sortButton"
          className="py-3 px-4 flex flex-row gap-2 items-center rounded-xl transition hover:bg-gray-200 hover:shadow-lg duration-300 ease-in-out"
          onClick={(event) => onCollapseSort(event)}
        >
          <span className="text-lg font-medium">Trier</span>
          <SortIcon />
        </button>
        {collapseSort && (
          <div ref={sortRef} className="absolute w-52 z-10 mt-14 bg-gray-100 shadow-lg rounded-xl p-4">
            <div className="flex flex-row"></div>
          </div>
        )}
      </div>
    </div>
  );
}
