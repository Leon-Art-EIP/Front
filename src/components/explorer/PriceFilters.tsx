import { useState } from "react";
import { IPriceRangeFilter, priceRangeFilters } from "../../configs/explorer/priceRanges";
import { IFilters } from "../../interfaces/explorer/filters";

interface PriceFiltersProps {
  handleFilters: (filters: IFilters) => void;
  filters: IFilters;
}

export default function PriceFilters(props: PriceFiltersProps): JSX.Element {
  const [priceRanges, setPriceRanges] = useState<IPriceRangeFilter[]>(priceRangeFilters);
  const [isNotForSale, setIsNotForSale] = useState<boolean>(false);

  function handleSelectPriceRange(priceRangeToggled: IPriceRangeFilter) {
    let clearPriceRange = false;

    priceRanges.forEach((priceRange) => {
      if (priceRange.priceRangeTitle === priceRangeToggled.priceRangeTitle) {
        if (priceRange.selected) clearPriceRange = true;
        priceRange.selected = !priceRange.selected;
      } else {
        priceRange.selected = false;
      }
    });

    if (!clearPriceRange) {
      setIsNotForSale(false);
    }
    props.handleFilters({
      ...props.filters,
      isForSale: false,
      priceRange: clearPriceRange ? "" : priceRangeToggled.priceRangeValue,
    });
  }

  function handleSelectIsForSale(isForSale: boolean) {
    setIsNotForSale(isForSale);
    priceRanges.forEach((priceRange) => {
      priceRange.selected = false;
    });
    props.handleFilters({ ...props.filters, priceRange: "", isForSale: isForSale });
  }

  return (
    <div className="flex flex-row gap-4">
      <button
        type="button"
        className={`py-2 w-full whitespace-nowrap border-secondary border-2 flex flex-row justify-between items-center px-4 rounded-xl hover:bg-secondary-hover hover:shadow-lg duration-300 ease-in-out text-tertiary ${
          isNotForSale ? "bg-secondary" : ""
        }`}
        onClick={() => handleSelectIsForSale(!isNotForSale)}
      >
        {"Indisponible Ã  la vente"}
      </button>
      {priceRanges.map((priceRange, index) => (
        <button
          key={index}
          type="button"
          className={`py-2 w-full whitespace-nowrap border-secondary border-2 flex flex-row justify-between items-center px-4 rounded-xl hover:bg-secondary-hover hover:shadow-lg duration-300 ease-in-out text-tertiary ${
            priceRange.selected ? "bg-gray-200" : ""
          }`}
          onClick={() => handleSelectPriceRange(priceRange)}
        >
          {priceRange.priceRangeTitle}
        </button>
      ))}
    </div>
  );
}
