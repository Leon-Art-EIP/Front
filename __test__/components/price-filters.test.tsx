import { test } from "vitest";
import { testSnapshot } from "../function/utils";
import { IFilters } from "../../src/interfaces/explorer/filters";
import PriceFilters from "../../src/components/explorer/PriceFilters";

const filters: IFilters = {
  searchTerm: "",
  artType: "",
  priceRange: "",
  isForSale: false,
  sort: "",
  artPage: 0,
  artLimit: 0,
  artistPage: 0,
  artistLimit: 0,
};

test("PriceFilters unavailableToBuy snapshot", () => {
  testSnapshot(
    <PriceFilters filters={filters} handleFilters={() => {}} setUnavailableToBuy={() => {}} unavailableToBuy />
  );
});

test("PriceFilters snapshot", () => {
  testSnapshot(
    <PriceFilters filters={filters} handleFilters={() => {}} setUnavailableToBuy={() => {}} unavailableToBuy={false} />
  );
});
