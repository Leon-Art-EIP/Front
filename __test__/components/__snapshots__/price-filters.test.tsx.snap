// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PriceFilters snapshot 1`] = `
<div
  className="flex flex-row gap-4"
>
  <button
    className="py-2 w-full whitespace-nowrap border-secondary border-2 flex flex-row justify-between items-center px-4 rounded-xl hover:bg-secondary-hover hover:shadow-lg duration-300 ease-in-out text-tertiary "
    onClick={[Function]}
    type="button"
  >
    0€ - 20€
  </button>
  <button
    className="py-2 w-full whitespace-nowrap border-secondary border-2 flex flex-row justify-between items-center px-4 rounded-xl hover:bg-secondary-hover hover:shadow-lg duration-300 ease-in-out text-tertiary "
    onClick={[Function]}
    type="button"
  >
    20€ - 50€
  </button>
  <button
    className="py-2 w-full whitespace-nowrap border-secondary border-2 flex flex-row justify-between items-center px-4 rounded-xl hover:bg-secondary-hover hover:shadow-lg duration-300 ease-in-out text-tertiary "
    onClick={[Function]}
    type="button"
  >
    50€ - 100€
  </button>
  <button
    className="py-2 w-full whitespace-nowrap border-secondary border-2 flex flex-row justify-between items-center px-4 rounded-xl hover:bg-secondary-hover hover:shadow-lg duration-300 ease-in-out text-tertiary "
    onClick={[Function]}
    type="button"
  >
    100€ - 500€
  </button>
  <button
    className="py-2 w-full whitespace-nowrap border-secondary border-2 flex flex-row justify-between items-center px-4 rounded-xl hover:bg-secondary-hover hover:shadow-lg duration-300 ease-in-out text-tertiary "
    onClick={[Function]}
    type="button"
  >
    + 500€
  </button>
</div>
`;

exports[`PriceFilters unavailableToBuy snapshot 1`] = `
<div
  className="flex flex-row gap-4"
>
  <button
    className="py-2 w-full whitespace-nowrap border-secondary border-2 flex flex-row justify-between items-center px-4 rounded-xl hover:bg-secondary-hover hover:shadow-lg duration-300 ease-in-out text-tertiary "
    onClick={[Function]}
    type="button"
  >
    0€ - 20€
  </button>
  <button
    className="py-2 w-full whitespace-nowrap border-secondary border-2 flex flex-row justify-between items-center px-4 rounded-xl hover:bg-secondary-hover hover:shadow-lg duration-300 ease-in-out text-tertiary "
    onClick={[Function]}
    type="button"
  >
    20€ - 50€
  </button>
  <button
    className="py-2 w-full whitespace-nowrap border-secondary border-2 flex flex-row justify-between items-center px-4 rounded-xl hover:bg-secondary-hover hover:shadow-lg duration-300 ease-in-out text-tertiary "
    onClick={[Function]}
    type="button"
  >
    50€ - 100€
  </button>
  <button
    className="py-2 w-full whitespace-nowrap border-secondary border-2 flex flex-row justify-between items-center px-4 rounded-xl hover:bg-secondary-hover hover:shadow-lg duration-300 ease-in-out text-tertiary "
    onClick={[Function]}
    type="button"
  >
    100€ - 500€
  </button>
  <button
    className="py-2 w-full whitespace-nowrap border-secondary border-2 flex flex-row justify-between items-center px-4 rounded-xl hover:bg-secondary-hover hover:shadow-lg duration-300 ease-in-out text-tertiary "
    onClick={[Function]}
    type="button"
  >
    + 500€
  </button>
</div>
`;
